integration/test/app/simple_test.go

```diff
@@ -0,0 +1,203 @@
+// +build k8srequired
+
+package app
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/e2e-harness/pkg/release"
+	"github.com/giantswarm/e2etemplates/pkg/chartvalues"
+	"github.com/giantswarm/microerror"
+	"golang.org/x/net/context"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/app-operator/integration/key"
+)
+
+const (
+	giantswarm = "giantswarm"
+)
+
+func TestAppLifecycle(t *testing.T) {
+	ctx := context.Background()
+
+	// Test creation.
+	{
+		config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("creating app %#q", key.CustomResourceReleaseName()))
+
+		c := chartvalues.APIExtensionsAppE2EConfig{
+			App: chartvalues.APIExtensionsAppE2EConfigApp{
+				Name:      key.TestAppReleaseName(),
+				Namespace: giantswarm,
+				Catalog:   key.TestAppCatalogReleaseName(),
+				Version:   "1.0.0",
+			},
+			AppCatalog: chartvalues.APIExtensionsAppE2EConfigAppCatalog{
+				Name:  key.TestAppCatalogReleaseName(),
+				Title: key.TestAppCatalogReleaseName(),
+				Storage: chartvalues.APIExtensionsAppE2EConfigAppCatalogStorage{
+					Type: "helm",
+					URL:  "https://giantswarm.github.com/sample-catalog",
+				},
+			},
+			AppOperator: chartvalues.APIExtensionsAppE2EConfigAppOperator{
+				Version: "1.0.0",
+			},
+			Namespace: giantswarm,
+		}
+
+		chartValues, err := chartvalues.NewAPIExtensionsAppE2E(c)
+		if err != nil {
+			t.Fatalf("expected %#v got %#v", nil, err)
+		}
+
+		chartInfo := release.NewStableChartInfo(key.CustomResourceReleaseName())
+		err = config.Release.Install(ctx, key.CustomResourceReleaseName(), chartInfo, chartValues)
+		if err != nil {
+			t.Fatalf("expected %#v got %#v", nil, err)
+		}
+
+		err = config.Release.WaitForStatus(ctx, fmt.Sprintf("%s-%s", giantswarm, key.CustomResourceReleaseName()), "DEPLOYED")
+		if err != nil {
+			t.Fatalf("expected %#v got %#v", nil, err)
+		}
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("created app %#q", key.CustomResourceReleaseName()))
+
+		config.Logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("checking chart CR %#q is deployed", key.TestAppReleaseName()))
+
+		tarballURL := "https://giantswarm.github.com/sample-catalog/test-app-1.0.0.tgz"
+		operation := func() error {
+			chart, err := config.Host.G8sClient().ApplicationV1alpha1().Charts(giantswarm).Get(key.TestAppReleaseName(), v1.GetOptions{})
+			if err != nil {
+				return microerror.Maskf(err, fmt.Sprintf("expected %#v got %#v", nil, err))
+			}
+			if !reflect.DeepEqual(chart.Spec.TarballURL, tarballURL) {
+				return microerror.Maskf(notMatching, fmt.Sprintf("expected tarballURL: %#v got %#v", tarballURL, chart.Spec.TarballURL))
```

Comment:
I'm not keen on the error handling here. I think it would be cleaner to get the chart CR with the backoff.

You can then check the CR spec and call test.Fatal if its wrong.

Reply:
This is creation test so it would work but what if it's update test?

Reply:
I think it still works. You need to get the chart CR using the g8s client. Update just needs to check the spec was modified correctly.

If the method getting the CR returns a notFound error it could also be used for delete. For delete you want the not found error. For create and update its an error case.

Reply:
We still could get the old `chart` CR in update test if use backoff function to get the CR, I think we will fail in some cases.

Reply:
OK I see what you mean. If it gets the chart CR too soon then the operator hasn't had a chance to apply the update.

One way around that is to check the resource version in ObjectMeta. k8s will increment that when the CR is updated.